File: scripts/auto-scrape-bbb.py
import re
import requests
from supabase import create_client
import os

# Init Supabase
supabase = create_client(os.environ['NEXT_PUBLIC_SUPABASE_URL'], os.environ['NEXT_PUBLIC_SUPABASE_ANON_KEY'])

def get_contractors_from_page(page_num):
    url = f"https://www.bbb.org/search?find_text=roofer&find_entity=10126-000&find_type=Category&find_loc=Salt+Lake+city%2C+UT&find_country=USA&page={page_num}"
    # Use requests to fetch (simpler for automated loops)
    response = requests.get(url, headers={'User-Agent': 'Mozilla/5.0'})
    
    # Simple regex extraction for Name/Phone pairs
    # Finding business names and phones in blocks
    names = re.findall(r'### \[(.*?)\]', response.text)
    phones = re.findall(r'\((\d{3}\)\s*\d{3}-\d{4})', response.text)
    
    extracted = []
    # Match by proximity or index (basic structure)
    for i in range(min(len(names), len(phones))):
        extracted.append({
            "business_name": names[i],
            "phone": '(' + phones[i],
            "source": "bbb-scrape-auto"
        })
    return extracted

def run_loop():
    for page in range(6, 16):
        print(f"Scraping page {page}...")
        contractors = get_contractors_from_page(page)
        if contractors:
            supabase.from_('contractors').insert(contractors).execute()
        print(f"Added {len(contractors)} contractors.")

run_loop()

File: scripts/auto-scrape-bbb.js
const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const https = require('https');

const envContent = fs.readFileSync('/data/.openclaw/workspace/prolink-ezly/.env', 'utf8');
const env = envContent.split('\n').reduce((acc, line) => {
    const [key, value] = line.split('=');
    if (key) acc[key.trim()] = value ? value.trim() : '';
    return acc;
}, {});
const supabase = createClient(env.NEXT_PUBLIC_SUPABASE_URL, env.NEXT_PUBLIC_SUPABASE_ANON_KEY);

const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));
const getRandomDelay = () => Math.floor(Math.random() * (15000 - 5000 + 1) + 5000);

function fetchPage(page) {
    return new Promise((resolve) => {
        const options = {
            hostname: 'www.bbb.org',
            path: '/search?find_text=roofer&find_entity=10126-000&find_type=Category&find_loc=Salt+Lake+city%2C+UT&find_country=USA&page=' + page,
            method: 'GET',
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
                'Referer': 'https://www.google.com/'
            }
        };

        https.get(options, (res) => {
            let data = '';
            res.on('data', (d) => data += d);
            res.on('end', () => resolve(data));
        }).on('error', (e) => resolve(''));
    });
}

function parse(html) {
    const contractors = [];
    const nameMatches = [...html.matchAll(/<h3><a.*?>(.*?)<\/a><\/h3>/gi)];
    const phoneMatches = [...html.matchAll(/\((\d{3})\)\s*(\d{3}-\d{4})/g)];
    
    for (let i = 0; i < Math.min(nameMatches.length, phoneMatches.length); i++) {
        contractors.push({
            business_name: nameMatches[i][1].replace(/<[^>]*>/g, '').trim(),
            phone: '(' + phoneMatches[i][1] + ') ' + phoneMatches[i][2],
            source: 'bbb-scrape-humanized'
        });
    }
    return contractors;
}

async function run() {
    for (let page = 6; page <= 15; page++) {
        console.log('--- Scraping page ' + page + ' with human mimicry ---');
        await sleep(getRandomDelay());
        const html = await fetchPage(page);
        const contractors = parse(html);
        if (contractors.length > 0) {
            await supabase.from('contractors').insert(contractors);
        }
        console.log('Inserted ' + contractors.length + ' contractors.');
    }
}
run();
